build.gradle

```diff
@@ -185,18 +187,76 @@ subprojects {
             // Jetty ALPN dependencies
             jetty_alpn_agent: 'org.mortbay.jetty.alpn:jetty-alpn-agent:2.0.9'
         ]
+
+        // A util function to config guava dependency with the given configuration
```

Comment:
Having each subproject list transitive dependency manipulation is easy to break by mistake as code evolves. For example, in one subproject there could be multiple dependencies excluding the same transitive dependencies. You either have to duplicate lines to add excluded items or scatter them in separate places.
```groovy
    implementation(libraries.guava) {
        // prefer our own versions from libraries instead of Guava's dependency
        exclude group: 'com.google.errorprone', module: 'error_prone_annotations'
        exclude group: 'com.google.code.findbugs', module: 'jsr305'
        exclude group: 'org.codehaus.mojo', module: 'animal-sniffer-annotations'
    }
    runtimeOnly libraries.jsr305
    implementation libraries.errorprone
    runtimeOnly libraries.animal-sniffer

    implementation (libraries.opencensus_api) {
        // prefer 3.0.2 from libraries instead of 3.0.1
        exclude group: 'com.google.code.findbugs', module: 'jsr305'
        // prefer 20.0 from libraries instead of 19.0
        exclude group: 'com.google.guava', module: 'guava'
        // we'll always be more up-to-date
        exclude group: 'io.grpc', module: 'grpc-context'
    }
    // duplicate line, otherwise you may forgot when other dependencies are removed.
    runtimeOnly libraries.jsr305
    // duplicate lines, otherwise you may forgot when other dependencies are removed.
    runtimeOnly(libraries.guava) {
        // prefer our own versions from libraries instead of Guava's dependency
        exclude group: 'com.google.errorprone', module: 'error_prone_annotations'
        exclude group: 'com.google.code.findbugs', module: 'jsr305'
        exclude group: 'org.codehaus.mojo', module: 'animal-sniffer-annotations'
    }
    // duplicate line, otherwise you may forgot when those in other dependencie are removed.
    runtimeOnly libraries.jsr305
    runtimeOnly libraries.errorprone
    runtimeOnly libraries.animal-sniffer
    runtimeOnly project(':grpc-context')

    implementation (libraries.opencensus_contrib_grpc_metrics) {
        // prefer 3.0.2 from libraries instead of 3.0.1
        exclude group: 'com.google.code.findbugs', module: 'jsr305'
        // we'll always be more up-to-date
        exclude group: 'io.grpc', module: 'grpc-context'
        // prefer 20.0 from libraries instead of 19.0
        exclude group: 'com.google.guava', module: 'guava'
    }
    // duplicate line, otherwise you may forgot when other dependencies are removed.
    runtimeOnly libraries.jsr305
    // duplicate lines, otherwise you may forgot when other dependencies are removed.
    runtimeOnly(libraries.guava) {
        // prefer our own versions from libraries instead of Guava's dependency
        exclude group: 'com.google.errorprone', module: 'error_prone_annotations'
        exclude group: 'com.google.code.findbugs', module: 'jsr305'
        exclude group: 'org.codehaus.mojo', module: 'animal-sniffer-annotations'
    }
    // duplicate line, otherwise you may forgot when those in other place are removed.
    runtimeOnly libraries.jsr305
    runtimeOnly libraries.errorprone
    runtimeOnly libraries.animal-sniffer
    runtimeOnly project(':grpc-context')
```

Reply:
I don't understand the example you posted (why Guava is added to implementation/runtimeOnly repeatedly?). But my point was what we currently have (not using utility functions) looks pretty straightforward as each subproject's buildscript clearly shows its dependencies, without redirecting to somewhere else to see what is included/excluded. Just kind of styling preference.

Reply:
>why Guava is added to implementation/runtimeOnly repeatedly?

The first Guava is added because the subproject needs it. The second Guava added is because it is excluded from libraries.opencensus_api you need added it back. If you don't added immediately after exclusion because it is already added somewhere els, then it is bug-prone as the build.gradle evolves.

Reply:
The `guava` dependency is an internal implementation detail, so it goes in the list of `implementation` deps. But that means it will not be automatically inherited to down-stream projects (which is what we want), including our own ("Oh.").

> having each subproject clearly list out its transitive dependency manipulation still seems to be more elegant, given that most users use a subset of subprojects.

`guavaDependency 'implementation'` is semantically equivalent to `implementation libraries.guava`. The difference is that the transitive dependencies versions don't match and `resolutionStrategy.failOnVersionConflict()` will fail the build. We have that check to avoid issues where Maven would choose the wrong version (since Maven chooses the first version encountered, not the greatest version number).

To "fix" `failOnVersionConflict` we have to exclude dependencies. But that is dangerous since we are excluding dependencies based on what's used _anywhere_ in gRPC, so a user may not get the dependency _at all_ if it is excluded. To make it reliable, we add the excluded dependencies explicitly. That makes it _much_ easier to audit, and removes all risk of things getting out-of-sync cross-module.

Note that `runtimeOnly libraries.jsr305` doesn't provide any "value" to a module, since it can't be used for compilation. So this utility is actually a pretty fair abstraction. The only place it breaks down is for error-prone with Guava, because the annotations must be available during compilation and so it is an `implementation` dependency.